### PR TITLE
Added Viewport canvas cull mask feature

### LINKED ISSUE #31856: Viewport: Unable to use parent world as world_2d
3.1.1.stable.official

Ubuntu 18.04

I want to have a viewport that renders the same scene as the main viewport but with a different camera. If I try to set the world_2d of the viewport as the main world_2d
vp2.world_2d =  get_viewport().world_2d
I get:
WARNING: set_world_2d: Unable to use parent world as world_2d
   At: scene/main/viewport.cpp:963.

And it doesn't work. 

However if I remove this check from the code everything works smoothly.

Is there any reason for this check?

### PR DESCRIPTION
**Pull request by BimDav**
Rewrite of https://github.com/godotengine/godot/pull/31877 for the master branch. Credits to @puthre.

There was no way to render canvasitems in specific viewports and this commit fixes that.
Added "Layers" property to canvas items and "Canvas Cull Mask" property to viewports so you can now enable / disable layers of canvas items per viewport.
Hopefully this will also help in adding effects to nodes (like glow) that need rendering
to texture.

It also fixes https://github.com/godotengine/godot/issues/31856. We can remove this check because the user can use layers and masks to prevent any issues that were previously prevented by this check

### CODE REVIEW

**Comment by Calinou** [comment]
This should also allow implementing https://github.com/godotengine/godot-proposals/issues/3218.

**Comment by BimDav** [comment]
> Does this PR close [godotengine/godot-proposals#2638](https://github.com/godotengine/godot-proposals/issues/2638) and/or [godotengine/godot-proposals#759](https://github.com/godotengine/godot-proposals/issues/759) by making them obsolete?

It seems so for the 2D part of these proposals. I am not sure concerning 3D ?

File: scene/main/canvas_item.cpp
Comment (reduz) [inline scene/main/canvas_item.cpp:1058]: The mask needs to be a uint32_t, so the high bit is not lost.

**Review by reduz — CHANGES_REQUESTED** [review]

**Comment by reduz** [comment]
This seems good to me, may be a bit of work to rebase because some files changed place though, but definitely wanted.

File: scene/main/viewport.cpp
Comment (clayjohn) [inline scene/main/viewport.cpp:1000]: Was this removed intentionally?

File: doc/classes/CanvasItem.xml
Comment (clayjohn) [inline doc/classes/CanvasItem.xml:464]: Docs will need to be filled in

File: servers/rendering/renderer_rd/renderer_canvas_render_rd.h
Comment (clayjohn) [inline servers/rendering/renderer_rd/renderer_canvas_render_rd.h:458]: I wonder, do we need to add ``canvas_cull_mask`` as a parameter of ``canvas_render_items`` if it is ultimately unused?

File: scene/main/viewport.cpp
Comment (BimDav) [reply scene/main/viewport.cpp:1000]: yes, it fixes this https://github.com/godotengine/godot/issues/31856, since you can use layers and masks to prevent any issues that were previously prevented by this check

File: doc/classes/CanvasItem.xml
Comment (BimDav) [reply doc/classes/CanvasItem.xml:464]: Ah, they were deleted in the property name change ><

File: servers/rendering/renderer_rd/renderer_canvas_render_rd.h
Comment (BimDav) [reply servers/rendering/renderer_rd/renderer_canvas_render_rd.h]: you are right, we no longer need it there

File: scene/main/viewport.cpp
Comment (Zireael07) [reply scene/main/viewport.cpp:1000]: probably worth a comment explaining this then

Comment (BimDav) [reply scene/main/viewport.cpp:1000]: Sorry, where should I comment?

Comment (clayjohn) [reply scene/main/viewport.cpp:1000]: Commenting in the PR description is fine

**Comment by BimDav** [comment]
I don't know what the problem is with the Linux build?

**Review by clayjohn — APPROVED** [review]
Looks good to me. @reduz will have to take another look before merging though

File: servers/rendering/renderer_viewport.h
Comment (reduz) [inline servers/rendering/renderer_viewport.h:120]: needs initialization to default value here

Comment (reduz) [reply servers/rendering/renderer_viewport.h:120]: To clarify, even if in other members they are not initialized due to legacy (pre C++11) code, for all new code we ask to add default initializers.

**Review by reduz — DISMISSED** [review]

**Comment by reduz** [comment]
Great job! Eager to finally merge this after changes and rebase.

**Comment by clayjohn** [comment]
Looks great! The final step before merging is to squash all the commits together so that the whole PR only contains 1 big commit with all your changes. We like to merge one commit at a time to keep the git history clean and navigable. 

If you don't know how to do that, we have a helpful tutorial in the official documentation https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase

**Comment by BimDav** [comment]
> We like to merge one commit at a time to keep the git history clean and navigable.

Ok, I thought there was something like squash and merge for PRs

**Comment by Zireael07** [comment]
@BimDav There is but Godot engine repo isn't using it

**Comment by BimDav** [comment]
Can I do something about the macOS build error?

**Comment by akien-mga** [comment]
To credit @puthre properly as co-author, you can add the following line at the end of the body of the commit message:
```
Co-authored-by: Valentin Zagura <puthre@gmail.com>
```
(The hyphens are important for it to be detected properly.)

**Pull request merged**
